Allow styling text in composer when selecting it with native actions

qfrank opened the pull request:
close(partially) #14137  

note: this PR doesn't contain corresponding implementation for Ios yet, someone's help would appreciated :)

qfrank committed:
Allow styling text in composer when selecting it with native actions

status-im-auto commented:
### Jenkins Builds
| :grey_question: | Commit | :hash: | Finished (UTC) | Duration | Platform | Result |
|-|-|-|-|-|-|-|
| :x: | c0fc6728 | [#2](https://ci.status.im/job/status-mobile/job/prs/job/tests/job/PR-14249/2/) | 2022-10-27 13:52:09 | ~4 min | `tests` | [:page_facing_up:`log`](https://ci.status.im/job/status-mobile/job/prs/job/tests/job/PR-14249/2/consoleText) |

ibrkhalil commented on src/status_im/ui2/screens/chat/composer/input.cljs:
```diff
@@ -168,12 +174,130 @@
-       (get @input-texts chat-id))]))
+     (fn []
+       (if mentions-enabled
+         (for [[idx [type text]] (map-indexed
```
We write indices as `index`

qfrank replied:
it's not my written code, but i agree..

OmarBasem commented:
Hey @qfrank, can you please explain why the implementation is done natively? As far as I understand from the designs it does not need to be styled in the composer itself.

qfrank commented:
> Hey @qfrank, can you please explain why the implementation is done natively? As far as I understand from the designs it does not need to be styled in the composer itself.

As someone else may have same question, i'd like to put this quote here:
"How could you add a custom menu item BIU to the popup menu without native implementation?"

ibrkhalil commented on modules/react-native-status/android/src/main/java/im/status/ethereum/module/RNSelectableTextInputViewManager.java:
```diff
@@ -0,0 +1,117 @@
+
+    @ReactProp(name = "menuItems")
+    public void setMenuItems(ReactViewGroup reactViewGroup, ReadableArray items) {
+        if(items != null) {
```
Can items be an empty array?

qfrank replied:
i'm not sure, but != null would prevent nullpointexception

ibrkhalil commented on modules/react-native-status/android/src/main/java/im/status/ethereum/module/RNSelectableTextInputViewManager.java:
```diff
@@ -0,0 +1,117 @@
+
+            this._menuItems = result.toArray(new String[items.size()]);
+
+            // PREVIOUS CODE
+            // registerSelectionListener(result.toArray(new String[items.size()]), reactViewGroup);
```
Should we keep dead code ?

qfrank replied:
yeah, we should remove

ibrkhalil commented on modules/react-native-status/android/src/main/java/im/status/ethereum/module/RNSelectableTextInputViewManager.java:
```diff
@@ -0,0 +1,117 @@
+package im.status.ethereum.module;
+
+
```
Unneeded spacing

qfrank replied:
okay

ibrkhalil commented on modules/react-native-status/android/src/main/java/im/status/ethereum/module/RNSelectableTextInputViewManager.java:
```diff
@@ -0,0 +1,117 @@
+import android.view.ActionMode.Callback;
+import android.view.Menu;
+import android.view.MenuItem;
+
```
Unneeded spacing

qfrank replied:
k

ibrkhalil commented on modules/react-native-status/android/src/main/java/im/status/ethereum/module/RNSelectableTextInputViewManager.java:
```diff
@@ -0,0 +1,117 @@
+import com.facebook.react.views.textinput.ReactEditText;
+import com.facebook.react.views.view.ReactViewGroup;
+import com.facebook.react.views.view.ReactViewManager;
+
```
Unneeded spacing

ibrkhalil commented on modules/react-native-status/android/src/main/java/im/status/ethereum/module/RNSelectableTextInputViewManager.java:
```diff
@@ -0,0 +1,117 @@
+    public void registerSelectionListener(final ReactEditText view) {
+        view.setCustomSelectionActionModeCallback(new Callback() {
+            @Override
+            public boolean onPrepareActionMode(ActionMode mode, Menu menu) {
+                // Called when action mode is first created. The menu supplied
+                // will be used to generate action buttons for the action mode
+                // Android Smart Linkify feature pushes extra options into the menu
+                // and would override the generated menu items
+                menu.clear();
+                for (int i = 0; i < _menuItems.length; i++) {
+                    menu.add(0, i, 0, _menuItems[i]);
+                }
+                return true;
```
Do we need this function to have a boolean return type if it's always returning true?
Can it be a `void` ?

qfrank replied:
it can't be void, as it's a interface implementation

ibrkhalil commented on modules/react-native-status/android/src/main/java/im/status/ethereum/module/RNSelectableTextInputViewManager.java:
```diff
@@ -0,0 +1,117 @@
+            }
+
+            @Override
+            public boolean onCreateActionMode(ActionMode mode, Menu menu) {
```
Do we need this function to have a boolean return type if it's always returning true?
Can it be a `void` ?

ibrkhalil commented on modules/react-native-status/android/src/main/java/im/status/ethereum/module/RNSelectableTextInputViewManager.java:
```diff
@@ -0,0 +1,117 @@
+
+            @Override
+            public void onDestroyActionMode(ActionMode mode) {
+                // Called when an action mode is about to be exited and
```
Can you complete this comment ?

ibrkhalil commented on modules/react-native-status/android/src/main/java/im/status/ethereum/module/RNSelectableTextInputViewManager.java:
```diff
@@ -0,0 +1,117 @@
+                         .put("topSelection", MapBuilder.of("registrationName","onSelection"))
+                         .build();
+    }
+}
```
Can you please add a new line at the end of this file ?

qfrank replied:
hi @ibrkhalil can u explain why we need a new line at the end of this file?

ibrkhalil replied:
Acts as a file terminator, And it makes our code base compatible with more tools
[Read](https://thoughtbot.com/blog/no-newline-at-end-of-file)

ibrkhalil commented on src/status_im/ui2/screens/chat/composer/input.cljs:
```diff
@@ -168,12 +174,130 @@
-         [rn/text (when (= type :mention) {:style {:color colors/primary-50}})
-          text])
-       (get @input-texts chat-id))]))
+     (fn []
```
Just curiosity, Why did we add this anonymous function ?

qfrank replied:
I need make this old snippet as callback

ibrkhalil commented on src/status_im/ui2/screens/chat/composer/input.cljs:
```diff
@@ -168,12 +174,130 @@
+
+(defn selectable-text-input-manager []
+  (when (exists? (.-NativeModules react-native))
+    (.-RNSelectableTextInputManager ^js (.-NativeModules react-native))))
```
Can this be written in a better way with `binaryage/oops`

qfrank replied:
how do u think? @flexsurfer

ibrkhalil commented on src/status_im/ui2/screens/chat/composer/input.cljs:
```diff
@@ -168,12 +174,130 @@
+  (when (exists? (.-NativeModules react-native))
+    (.-RNSelectableTextInputManager ^js (.-NativeModules react-native))))
+
+(defonce rn-selectable-text-input (reagent/adapt-react-class (.requireNativeComponent react-native "RNSelectableTextInput")))
```
Can we use `binaryage/oops`

ibrkhalil commented on src/status_im/ui2/screens/chat/composer/input.cljs:
```diff
@@ -168,12 +174,130 @@
+        tail (subs full-text selection-end)]
+    (str head content tail)))
+
+(def first-level-menus {:cut (fn [{:keys [content] :as params}]
+                               (let [new-text (calculate-input-text params "")]
+                                 (react/copy-to-clipboard content)
+                                 (update-input-text params new-text)))
+
+                        :copy-to-clipboard (fn [{:keys [content]}]
+                                             (react/copy-to-clipboard content))
+
+                        :paste (fn [params]
```
Since all operations happens on the clipboard, Can we rename the key `:copy-to-clipboard` to `:copy``

qfrank replied:
i don't think so, as we may use this keyword relate to i18n

ibrkhalil commented on src/status_im/ui2/screens/chat/composer/input.cljs:
```diff
@@ -168,12 +174,130 @@
+                               (let [show-menu #(.startActionMode (selectable-text-input-manager) text-input-handle)]
+                                 (reset! first-level false)
+                                 (reset! menu-items second-level-menu-items)
+                                 ;to avoid something discusting like this https://lightrun.com/answers/facebook-react-native-textinput-controlled-selection-broken-on-both-ios-and-android
```
Disgusting*

qfrank replied:
lol, sorry for my spell

ibrkhalil commented on src/status_im/ui2/screens/chat/composer/input.cljs:
```diff
@@ -168,12 +174,130 @@
+                                 (reset! menu-items second-level-menu-items)
+                                 ;to avoid something discusting like this https://lightrun.com/answers/facebook-react-native-textinput-controlled-selection-broken-on-both-ios-and-android
+                                 ;use native invoke instead! do not use setNativeProps! e.g. (.setNativeProps ^js text-input (clj->js {:selection {:start selection-start :end selection-end}}))
+                                 (.setSelection (selectable-text-input-manager) text-input-handle selection-start selection-end)
```
Can use `->` for readability ?

ibrkhalil commented on src/status_im/ui2/screens/chat/composer/input.cljs:
```diff
@@ -168,12 +174,130 @@
+                                 ;use native invoke instead! do not use setNativeProps! e.g. (.setNativeProps ^js text-input (clj->js {:selection {:start selection-start :end selection-end}}))
+                                 (.setSelection (selectable-text-input-manager) text-input-handle selection-start selection-end)
+                                 ;set a delay to wait RNSelectableTextInputViewManager#_menuItems get updated, need figure out a better way
+                                 (js/setTimeout show-menu 200)))})
```
Have you tried React's lifecycle methods for this ?

ibrkhalil commented on src/status_im/ui2/screens/chat/composer/input.cljs:
```diff
@@ -168,12 +174,130 @@
+                                 ;set a delay to wait RNSelectableTextInputViewManager#_menuItems get updated, need figure out a better way
+                                 (js/setTimeout show-menu 200)))})
+
+(def first-level-menu-items (map i18n/label (keys first-level-menus)))
```
Instead of calling keys to iterate over map keys, You can do
`(map (fn [[k _]] (i18n/label k)) first-level-menus)`

ibrkhalil replied:
Instead of calling keys to iterate over map keys, You can do
`(map (fn [[k _]] (i18n/label k)) first-level-menus)`

ibrkhalil commented on src/status_im/ui2/screens/chat/composer/input.cljs:
```diff
@@ -168,12 +174,130 @@
+  (reset! menu-items first-level-menu-items))
+
+(defn append-markdown-char [{:keys [first-level menu-items content] :as params} wrap-chars]
+  (let [content (str wrap-chars content wrap-chars)
+        new-text (calculate-input-text params content)]
+    (update-input-text params new-text)
+    (reset-to-first-level-menu first-level menu-items)))
```
Align let plz

ibrkhalil commented on src/status_im/ui2/screens/chat/composer/input.cljs:
```diff
@@ -168,12 +174,130 @@
+                                   (react/get-from-clipboard callback)))
+
+                        :biu (fn [{:keys [first-level text-input-handle menu-items selection-start selection-end]}]
+                               (let [show-menu #(.startActionMode (selectable-text-input-manager) text-input-handle)]
```
Can use `->` for readability ?

ibrkhalil commented on src/status_im/ui2/screens/chat/composer/input.cljs:
```diff
@@ -168,12 +174,130 @@
+    (update-input-text params new-text)
+    (reset-to-first-level-menu first-level menu-items)))
+
+(def second-level-menus {:bold #(append-markdown-char % "**")
+
+                         :italic #(append-markdown-char % "*")
+
+                         :strikethrough #(append-markdown-char % "~~")})
```
Kindly correct me if am wrong ..
I see a `:biu` keyword in the code which I assume it stands for bold/italic/underline
But I can't seem to find an underlining option anywhere ..

qfrank replied:
it comes from the design

ibrkhalil commented on src/status_im/ui2/screens/chat/composer/input.cljs:
```diff
@@ -168,12 +174,130 @@
+
+(def second-level-menu-items (map i18n/label (keys second-level-menus)))
+
+(defn on-menu-item-touched [{:keys [first-level event-type] :as params}]
+  (let [menus (if @first-level first-level-menus second-level-menus)
+        menu-item-key (nth (keys menus) event-type)
+        action (get menus menu-item-key)]
+    (action params)))
```
Align let please

ibrkhalil commented on src/status_im/ui2/screens/chat/composer/input.cljs:
```diff
@@ -168,12 +174,130 @@
+        menu-items (reagent/atom first-level-menu-items)
+        first-level (reagent/atom true)]
+    (reagent/create-class
+     {:component-did-mount (fn [self]
+                             (when @text-input-ref
+                               (let [self-handle (rn/find-node-handle self)
```
We usually use `this`

OmarBasem replied:
Can this be a functional component instead of a class component?

qfrank replied:
Hi @OmarBasem , i would prefer keep the implementation of class component to functional component ATM as u know there would cause new issue (can't type in the message input) after using functional component :) , @flexsurfer agreed

ibrkhalil commented on src/status_im/ui2/screens/chat/composer/input.cljs:
```diff
@@ -168,12 +174,130 @@
+                                 (.setupMenuItems (selectable-text-input-manager) self-handle text-input-handle))))
+      :render
+      (fn [_]
+        (let [old-ref (:ref props)
+              ref #(do (reset! text-input-ref %)
+                       (when old-ref
+                         (quo.react/set-ref-val! old-ref %)))
+              old-on-selection-change (:on-selection-change props)
+              on-selection-change (fn [args]
+                                    (let [selection (.-selection ^js (.-nativeEvent ^js args))
+                                          start (.-start selection)
+                                          end (.-end selection)
+                                          no-selection (<= (- end start) 0)]
+                                      (when no-selection
+                                        (do (.hideLastActionMode (selectable-text-input-manager))
+                                            (reset-to-first-level-menu first-level menu-items))))
+                                    (when old-on-selection-change
+                                      (old-on-selection-change args)))
+              props (merge props {:ref ref
+                                  :style nil
```
Align `let` and the `map` please

ibrkhalil commented on src/status_im/ui2/screens/chat/composer/input.cljs:
```diff
@@ -168,12 +174,130 @@
+                                          end (.-end selection)
+                                          no-selection (<= (- end start) 0)]
+                                      (when no-selection
+                                        (do (.hideLastActionMode (selectable-text-input-manager))
+                                            (reset-to-first-level-menu first-level menu-items))))
```
```clj
(-> (selectable-text-input-manager)
     .hideLastActionMode)
```
is more readable

ibrkhalil commented on src/status_im/ui2/screens/chat/composer/input.cljs:
```diff
@@ -168,12 +174,130 @@
+                                                                           :chat-id chat-id})))})]
+          [rn-selectable-text-input {:menuItems @menu-items :style style}
+           [rn/text-input props
+            [children]]]))})))
```
Can you please add a new line at the end of this file ?

ibrkhalil reviewed (changes requested):
Is there a way to add unit testing to this functionality ?

OmarBasem commented on src/status_im/ui2/screens/chat/composer/input.cljs:
```diff
@@ -168,12 +174,130 @@
+         (for [[idx [type text]] (map-indexed
+                                  (fn [idx item]
+                                    [idx item])
+                                  (<sub [:chat/input-with-mentions]))]
```
Please change `<sub` to `rf/sub` in this PR and future PRs. Also, for dispatching events we will be using `rf/dispatch`.

OmarBasem commented on src/status_im/ui2/screens/chat/composer/input.cljs:
```diff
@@ -168,12 +174,130 @@
+                                                  (let [native-event (.-nativeEvent event)
+                                                        native-event (types/js->clj native-event)
+                                                        {:keys [eventType content selectionStart selectionEnd]} native-event
+                                                        full-text (:input-text (<sub [:chats/current-chat-inputs]))]
```
Here also, change `<sub` to `rf/sub`

ilmotta commented:
@qfrank, I tried to run the app locally, but unfortunately https://github.com/qfrank/status-react/tree/feature/14137 fails to me with the error below. Even the develop branch in your fork fails to me. To double-check it wasn't a problem on my machine, I rebased `upstream/develop` onto `origin/develop` (locally) and everything worked just fine. So, could you please update the fork with the latest changes from `status-im/status-mobile` `develop` branch, so that I can test your PR?

```
node_modules/react-native-blob-util/android/src/main/java/com/ReactNativeBlobUtil/ReactNativeBlobUtilUtils.java:3: error: package com.facebook.react.bridge does not exist
import com.facebook.react.bridge.Arguments;
```

qfrank commented:
> @qfrank, I tried to run the app locally, but unfortunately https://github.com/qfrank/status-react/tree/feature/14137 fails to me with the error below. Even the develop branch in your fork fails to me. To double-check it wasn't a problem on my machine, I rebased `upstream/develop` onto `origin/develop` (locally) and everything worked just fine. So, could you please update the fork with the latest changes from `status-im/status-mobile` `develop` branch, so that I can test your PR?
> 
> ```
> node_modules/react-native-blob-util/android/src/main/java/com/ReactNativeBlobUtil/ReactNativeBlobUtilUtils.java:3: error: package com.facebook.react.bridge does not exist
> import com.facebook.react.bridge.Arguments;
> ```

Hi @ilmotta , rebased, enjoy!

qfrank commented:
Hi @ibrkhalil , thanks for your review, a lot of align task :), can we leave the align task later to `make lint-fix`? currently, i'm using IntelliJ with cursive plugin, i've made an auto format with IDE, then `make lint-fix`, but `make lint-fix` won't align let or map ATM, i hope we can get this [pr](https://github.com/status-im/status-mobile/pull/14128) merged ASAP. @flexsurfer @cammellos thoughts?

qfrank commented:
> Is there a way to add unit testing to this functionality ?

i don't think we need here, if u know there is a way, i'd like to hear

ilmotta commented:
> Hi @ilmotta , rebased, enjoy!

Thanks @qfrank. I could successfully run it now.

So I tried in the Android Emulator, and I took two screenshots (below) where we can see the native BIU menu option. The functionality works, but the behavior and implementation is completely different from what I see in Figma (https://www.figma.com/file/wA8Epdki2OWa8Vr067PCNQ/Composer-for-Mobile?node-id=2230%3A116813). A couple important differences:

- The Figma spec clearly shows the styling should not be applied on selection, i.e. the user is not required to select the text.
- In the Figma file, I can see the **styling BIU buttons are not native**. We can even see button icons are part of our design system.

<table>
  <thead>
    <tr>
      <th>
        <img height="400" alt="step-01" src="https://user-images.githubusercontent.com/46027/200794035-4cf1ee5a-55cc-4efd-b7de-03e9d3cd0cf3.png"/>
      </th>
      <th>
        <img height="400" alt="step-02" src="https://user-images.githubusercontent.com/46027/200794049-d4ee2744-efde-4b88-b9df-0733c2dc6d21.png"/>
      </th>
    </tr>
  </thead>
</table>

> "How could you add a custom menu item BIU to the popup menu without native implementation?"

Where's the Figma spec showing we should implement a native BIU option in the popup menu? The image below shows the styling button is not in a popup menu.

<img src="https://user-images.githubusercontent.com/46027/200797398-e8b1478c-80a0-4be2-8020-bfd2c7ba75e7.png" height="200" />

Hence I actually don't understand how this PR is implementing what the Figma specs require. Could you help me understand why the mismatch between this PR's implementation and the Figma specs?

ilmotta commented:
> Hi @ibrkhalil , thanks for your review, a lot of align task :), can we leave the align task later to `make lint-fix`? currently, i'm using IntelliJ with cursive plugin, i've made an auto format with IDE, then `make lint-fix`, but `make lint-fix` won't align let or map ATM, i hope we can get this [pr](https://github.com/status-im/status-mobile/pull/14128) merged ASAP. @flexsurfer @cammellos thoughts?

I don't know if it helps you, but this Cursive page says aligning let bindings is supported.
https://cursive-ide.com/userguide/formatting.html

And you're right, `make lint-fix` never supported aligning map keys and let bindings, so all of us are using what our IDEs/editors support. If/when we start using zprint, then this tool will actually be able to format map keys and let bindings.

qfrank commented:
> > Hi @ilmotta , rebased, enjoy!
> 
> Thanks @qfrank. I could successfully run it now.
> 
> So I tried in the Android Emulator, and I took two screenshots (below) where we can see the native BIU menu option. The functionality works, but the behavior and implementation is completely different from what I see in Figma (https://www.figma.com/file/wA8Epdki2OWa8Vr067PCNQ/Composer-for-Mobile?node-id=2230%3A116813). A couple important differences:
> 
> * The Figma spec clearly shows the styling should not be applied on selection, i.e. the user is not required to select the text.
> * In the Figma file, I can see the **styling BIU buttons are not native**. We can even see button icons are part of our design system.
> 
> <img alt="step-01" height="400" src="https://user-images.githubusercontent.com/46027/200794035-4cf1ee5a-55cc-4efd-b7de-03e9d3cd0cf3.png">	<img alt="step-02" height="400" src="https://user-images.githubusercontent.com/46027/200794049-d4ee2744-efde-4b88-b9df-0733c2dc6d21.png">
> > "How could you add a custom menu item BIU to the popup menu without native implementation?"
> 
> Where's the Figma spec showing we should implement a native BIU option in the popup menu? The image below shows the styling button is not in a popup menu.
> 
> <img alt="" height="200" src="https://user-images.githubusercontent.com/46027/200797398-e8b1478c-80a0-4be2-8020-bfd2c7ba75e7.png">
> 
> Hence I actually don't understand how this PR is implementing what the Figma specs require. Could you help me understand why the mismatch between this PR's implementation and the Figma specs?

Good questions, i had a conversation with petro.et before (i should ask it in public channel ):), let me quote here:
> me : when the text get selected, the menu popup which contains menu items like Copy/Paste../BIU etc. Could u explain why we need duplicate it as we have our own in the bottom ? [snapshot](https://cdn.discordapp.com/attachments/1027845131761635349/1032499248190730300/unknown.png)
petro: so in that particular instance the formatting options are open but it may be that the user selects text when its not. so at the bottom there would be the add image, add reactions, format buttons [snapshot](https://cdn.discordapp.com/attachments/1032568672218271765/1032588076955992104/Screenshot_2022-10-20_at_10.36.17.png) , so if the user would select text at this point in time

I guess above would answer the first difference? 
On the second difference, i've discussed with @cammellos  before that i don't have to implement the bottom part ATM. it would be a separate issue.
Also the style of popup menu would be a separate one too. @ilmotta

qfrank commented:
> > Hi @ibrkhalil , thanks for your review, a lot of align task :), can we leave the align task later to `make lint-fix`? currently, i'm using IntelliJ with cursive plugin, i've made an auto format with IDE, then `make lint-fix`, but `make lint-fix` won't align let or map ATM, i hope we can get this [pr](https://github.com/status-im/status-mobile/pull/14128) merged ASAP. @flexsurfer @cammellos thoughts?
> 
> I don't know if it helps you, but this Cursive page says aligning let bindings is supported. https://cursive-ide.com/userguide/formatting.html
> 
> And you're right, `make lint-fix` never supported aligning map keys and let bindings, so all of us are using what our IDEs/editors support. If/when we start using zprint, then this tool will actually be able to format map keys and let bindings.

U saved my whole day! thanks a lot @ilmotta

ilmotta commented:
> I guess above would answer the first difference? On the second difference, i've discussed with @cammellos before that i don't have to implement the bottom part ATM. it would be a separate issue. Also the style of popup menu would be a separate one too. @ilmotta

Thanks for the clarifications @qfrank, it seems we're good for now :)

Changing subject:

Figma shows a menu option with styling, can we do that in the native menu option for Android? It would be great if we could do that, because the acronym `BIU` without styling feels weird and I'd guess 98% of users don't know what `BIU` stands for (I didn't tbh).

<img width="400" src="https://user-images.githubusercontent.com/46027/200822862-49d7df79-cdd6-445b-92f7-034e40ae472e.png" />

qfrank commented:
> > I guess above would answer the first difference? On the second difference, i've discussed with @cammellos before that i don't have to implement the bottom part ATM. it would be a separate issue. Also the style of popup menu would be a separate one too. @ilmotta
> 
> Thanks for the clarifications @qfrank, it seems we're good for now :)
> 
> Changing subject:
> 
> Figma shows a menu option with styling, can we do that in the native menu option for Android? It would be great if we could do that, because the acronym `BIU` without styling feels weird and I'd guess 98% of users don't know what `BIU` stands for (I didn't tbh).
> 
> <img alt="" width="400" src="https://user-images.githubusercontent.com/46027/200822862-49d7df79-cdd6-445b-92f7-034e40ae472e.png">

i believe we can style the popup menu, but it's another issue we need process :)

ilmotta commented:
> i believe we can style the popup menu, but it's another issue we need process :)

Sounds great to me. On top of that, in the PR description you mentioned about getting help from someone with iOS experience. I guess that could be another issue as well, right?

qfrank commented:
i'd like to hear something from @cammellos , as he said he knows OC as well

cammellos commented:
> i'd like to hear something from @cammellos , as he said he knows OC as well

nope, pretty sure I didn't say that ;)

I probably meant that I know some people who do :)

@qfrank would you mind making sure once it's merged that the issue is not closed and it's updated to say that we need the IOS implementation?
Thank you!

cammellos commented:
> i don't suggest merge it right now, as it may impact someone else who develop on ios, since we don't have the ios implementation yet. Maybe we can wait for the ios implementation, then get merged together? @cammellos

If it works, and does not break IOS, I think it's fine to merge if we are happy with it, whoever picks up IOS will use it as a reference, and we don't have to rebase etc

btw linting is failing

qfrank commented:
> > i don't suggest merge it right now, as it may impact someone else who develop on ios, since we don't have the ios implementation yet. Maybe we can wait for the ios implementation, then get merged together? @cammellos
> 
> If it works, and does not break IOS, I think it's fine to merge if we are happy with it, whoever picks up IOS will use it as a reference, and we don't have to rebase etc
> 
> btw linting is failing

i'm afraid it will break IOS ATM, like this 
```
ERROR    Invariant Violation: requireNativeComponent: "RNSelectableTextInput" was not found in the UIManager.
``` , need more work to make it not break, agree more work?

yeah, i've noticed the lint issue.

status-im-auto commented:
## 33% of end-end tests have passed
```
Total executed tests: 6
Failed tests: 4
Passed tests: 2
Not executed tests: 1
```
```
IDs of not executed tests: 702777 
```
```
IDs of failed tests: 702733,702776,702745,702730 
```
<h3>Not executed tests (1)</h3><details><summary>Click to expand</summary><br/><li><a href="https://ci.status.im/job/status-mobile/job/e2e/job/status-app-prs-rerun/parambuild/?BRANCH_NAME=develop&APK_NAME=14249.apk&PR_ID=14249&TR_CASE_IDS=702777">Rerun not executed tests</a></li></details><h3>Failed tests (4)</h3><details><summary>Click to expand</summary><br/><li><a href="https://ci.status.im/job/status-mobile/job/e2e/job/status-app-prs-rerun/parambuild/?BRANCH_NAME=develop&APK_NAME=14249.apk&PR_ID=14249&TR_CASE_IDS=702733,702776,702745,702730">Rerun failed tests</a></li><br/><h4>Class TestOneToOneChatMultipleSharedDevicesNewUi:</h4><table style="width: 100%"><colgroup><col span="1" style="width: 20%;"><col span="1" style="width: 80%;"></colgroup><tbody><tr></tr><tr><td><b>1. <a href="https://ethstatus.testrail.net/index.php?/tests/view/3046987">test_1_1_chat_text_message_delete_push_disappear</a>, id: 702733 </b></td></tr><tr><td>

```
Test setup failed: critical/chats/test_1_1_public_chats.py:1235: in prepare_devices
    self.chat_1.send_message('hey')
../views/chat_view.py:826: in send_message
    self.chat_message_input.wait_for_element(wait_chat_input_sec)
../views/base_element.py:116: in wait_for_element
    raise TimeoutException(
 Device `1`: `EditBox` by` accessibility id`: `chat-message-input` is not found on the screen after wait_for_element
```

<br/><br/></td></tr><tr><td><b>2. <a href="https://ethstatus.testrail.net/index.php?/tests/view/3046988">test_1_1_chat_non_latin_messages_stack_update_profile_photo</a>, id: 702745 </b></td></tr><tr><td>

```
Test setup failed: critical/chats/test_1_1_public_chats.py:1235: in prepare_devices
    self.chat_1.send_message('hey')
../views/chat_view.py:826: in send_message
    self.chat_message_input.wait_for_element(wait_chat_input_sec)
../views/base_element.py:116: in wait_for_element
    raise TimeoutException(
 Device `1`: `EditBox` by` accessibility id`: `chat-message-input` is not found on the screen after wait_for_element
```

<br/><br/></td></tr><tr><td><b>3. <a href="https://ethstatus.testrail.net/index.php?/tests/view/3046986">test_1_1_chat_message_reaction</a>, id: 702730 </b></td></tr><tr><td><p><blockquote><div>## 1-1 chat is created successfully!</div><div>Device 1: Sending message 'hey' </div></blockquote></p>

```
Test setup failed: critical/chats/test_1_1_public_chats.py:1235: in prepare_devices
    self.chat_1.send_message('hey')
../views/chat_view.py:826: in send_message
    self.chat_message_input.wait_for_element(wait_chat_input_sec)
../views/base_element.py:116: in wait_for_element
    raise TimeoutException(
 Device `1`: `EditBox` by` accessibility id`: `chat-message-input` is not found on the screen after wait_for_element
```

<br/><br/><ins>Device sessions</ins><p><ul><p>Device 1:<ul><li><a href="https://eu-central-1.saucelabs.com/jobs/078676d36ad24746b7a0856fb4478a26?auth=c5eb8d99ad5bc500490a7f0cc3ba49ea">Steps, video, logs</a></li><li><a href="https://assets.eu-central-1.saucelabs.com/jobs/078676d36ad24746b7a0856fb4478a26/0019screenshot.png?auth=c5eb8d99ad5bc500490a7f0cc3ba49ea">Failure screenshot</a></li></ul></p><p>Device 2:<ul><li><a href="https://eu-central-1.saucelabs.com/jobs/eeb86efc36eb4e88bfd130a62f1689ba?auth=bb14ea23f28379c000581721c8fa37ec">Steps, video, logs</a></li><li><a href="https://assets.eu-central-1.saucelabs.com/jobs/eeb86efc36eb4e88bfd130a62f1689ba/0011screenshot.png?auth=bb14ea23f28379c000581721c8fa37ec">Failure screenshot</a></li></ul></p></ul></p></td></tr></tbody></table><h4>Class TestDeeplinkOneDevice:</h4><table style="width: 100%"><colgroup><col span="1" style="width: 20%;"><col span="1" style="width: 80%;"></colgroup><tbody><tr></tr><tr><td><b>1. <a href="https://ethstatus.testrail.net/index.php?/tests/view/3046991">test_public_chat_open_using_deep_link</a>, id: 702776 </b></td></tr><tr><td><p><blockquote><div>## Signed in successfully!</div><div>Device 1: Find `Text` by `xpath`: `//android.view.ViewGroup/android.widget.TextView` </div></blockquote></p>

```
medium/test_deeplink_chat_share_profile.py:380: in test_public_chat_open_using_deep_link
    if not chat.user_name_text_new_UI.text == '#' + chat_name:
../views/base_element.py:357: in text
    text = self.find_element().text
../views/base_element.py:79: in find_element
    raise NoSuchElementException(
 Device 1: Text by xpath: `//android.view.ViewGroup/android.widget.TextView` is not found on the screen; also Unexpected Alert is shown: 'undefined is not a function'
```

<br/><br/><ins>Device sessions</ins><p><ul><p>Device 1:<ul><li><a href="https://eu-central-1.saucelabs.com/jobs/78f5bfef14c24b49829e3dc116407246?auth=3d8b4ff1a3dd13fd75bb119094bb04f5#79">Steps, video, logs</a></li><li><a href="https://assets.eu-central-1.saucelabs.com/jobs/78f5bfef14c24b49829e3dc116407246/0029screenshot.png?auth=3d8b4ff1a3dd13fd75bb119094bb04f5">Failure screenshot</a></li></ul></p></ul></p></td></tr></tbody></table></details><h3>Passed tests (2)</h3><details><summary>Click to expand</summary><br/><br/><h4>Class TestDeeplinkOneDevice:</h4><table style="width: 100%"><colgroup><col span="1" style="width: 20%;"><col span="1" style="width: 80%;"></colgroup><tbody><tr></tr><tr><td><b>1. <a href="https://ethstatus.testrail.net/index.php?/tests/view/3046989">test_deep_link_with_invalid_user_public_key_own_profile_key</a>, id: 702774 </b></td></tr><tr><td><ins>Device sessions</ins><p><ul><p>Device 1:<ul><li><a href="https://eu-central-1.saucelabs.com/jobs/78f5bfef14c24b49829e3dc116407246?auth=3d8b4ff1a3dd13fd75bb119094bb04f5#40">Steps, video, logs</a></li></ul></p></ul></p></td></tr><tr><td><b>2. <a href="https://ethstatus.testrail.net/index.php?/tests/view/3046990">test_deep_link_open_user_profile</a>, id: 702775 </b></td></tr><tr><td><ins>Device sessions</ins><p><ul><p>Device 1:<ul><li><a href="https://eu-central-1.saucelabs.com/jobs/78f5bfef14c24b49829e3dc116407246?auth=3d8b4ff1a3dd13fd75bb119094bb04f5#98">Steps, video, logs</a></li></ul></p></ul></p></td></tr></tbody></table></details>

status-im-auto commented:
## 33% of end-end tests have passed
```
Total executed tests: 6
Failed tests: 4
Passed tests: 2
Not executed tests: 1
```
```
IDs of not executed tests: 702777 
```
```
IDs of failed tests: 702733,702776,702745,702730 
```
<h3>Not executed tests (1)</h3><details><summary>Click to expand</summary><br/><li><a href="https://ci.status.im/job/status-mobile/job/e2e/job/status-app-prs-rerun/parambuild/?BRANCH_NAME=develop&APK_NAME=14249.apk&PR_ID=14249&TR_CASE_IDS=702777">Rerun not executed tests</a></li></details><h3>Failed tests (4)</h3><details><summary>Click to expand</summary><br/><li><a href="https://ci.status.im/job/status-mobile/job/e2e/job/status-app-prs-rerun/parambuild/?BRANCH_NAME=develop&APK_NAME=14249.apk&PR_ID=14249&TR_CASE_IDS=702733,702776,702745,702730">Rerun failed tests</a></li><br/><h4>Class TestDeeplinkOneDevice:</h4><table style="width: 100%"><colgroup><col span="1" style="width: 20%;"><col span="1" style="width: 80%;"></colgroup><tbody><tr></tr><tr><td><b>1. <a href="https://ethstatus.testrail.net/index.php?/tests/view/3047168">test_public_chat_open_using_deep_link</a>, id: 702776 </b></td></tr><tr><td><p><blockquote><div>## Signed in successfully!</div><div>Device 1: Find `Text` by `xpath`: `//android.view.ViewGroup/android.widget.TextView` </div></blockquote></p>

```
medium/test_deeplink_chat_share_profile.py:380: in test_public_chat_open_using_deep_link
    if not chat.user_name_text_new_UI.text == '#' + chat_name:
../views/base_element.py:357: in text
    text = self.find_element().text
../views/base_element.py:79: in find_element
    raise NoSuchElementException(
 Device 1: Text by xpath: `//android.view.ViewGroup/android.widget.TextView` is not found on the screen; also Unexpected Alert is shown: 'undefined is not a function'
```

<br/><br/><ins>Device sessions</ins><p><ul><p>Device 1:<ul><li><a href="https://eu-central-1.saucelabs.com/jobs/11e17eb0207d44328cd7a142a18f8ab8?auth=9a43fdd0b4078a224b2f85ad76c1d855#79">Steps, video, logs</a></li><li><a href="https://assets.eu-central-1.saucelabs.com/jobs/11e17eb0207d44328cd7a142a18f8ab8/0029screenshot.png?auth=9a43fdd0b4078a224b2f85ad76c1d855">Failure screenshot</a></li></ul></p></ul></p></td></tr></tbody></table><h4>Class TestOneToOneChatMultipleSharedDevicesNewUi:</h4><table style="width: 100%"><colgroup><col span="1" style="width: 20%;"><col span="1" style="width: 80%;"></colgroup><tbody><tr></tr><tr><td><b>1. <a href="https://ethstatus.testrail.net/index.php?/tests/view/3047164">test_1_1_chat_text_message_delete_push_disappear</a>, id: 702733 </b></td></tr><tr><td>

```
Test setup failed: critical/chats/test_1_1_public_chats.py:1235: in prepare_devices
    self.chat_1.send_message('hey')
../views/chat_view.py:826: in send_message
    self.chat_message_input.wait_for_element(wait_chat_input_sec)
../views/base_element.py:116: in wait_for_element
    raise TimeoutException(
 Device `1`: `EditBox` by` accessibility id`: `chat-message-input` is not found on the screen after wait_for_element
```

<br/><br/></td></tr><tr><td><b>2. <a href="https://ethstatus.testrail.net/index.php?/tests/view/3047165">test_1_1_chat_non_latin_messages_stack_update_profile_photo</a>, id: 702745 </b></td></tr><tr><td>

```
Test setup failed: critical/chats/test_1_1_public_chats.py:1235: in prepare_devices
    self.chat_1.send_message('hey')
../views/chat_view.py:826: in send_message
    self.chat_message_input.wait_for_element(wait_chat_input_sec)
../views/base_element.py:116: in wait_for_element
    raise TimeoutException(
 Device `1`: `EditBox` by` accessibility id`: `chat-message-input` is not found on the screen after wait_for_element
```

<br/><br/></td></tr><tr><td><b>3. <a href="https://ethstatus.testrail.net/index.php?/tests/view/3047163">test_1_1_chat_message_reaction</a>, id: 702730 </b></td></tr><tr><td><p><blockquote><div>## 1-1 chat is created successfully!</div><div>Device 1: Sending message 'hey' </div></blockquote></p>

```
Test setup failed: critical/chats/test_1_1_public_chats.py:1235: in prepare_devices
    self.chat_1.send_message('hey')
../views/chat_view.py:826: in send_message
    self.chat_message_input.wait_for_element(wait_chat_input_sec)
../views/base_element.py:116: in wait_for_element
    raise TimeoutException(
 Device `1`: `EditBox` by` accessibility id`: `chat-message-input` is not found on the screen after wait_for_element
```

<br/><br/><ins>Device sessions</ins><p><ul><p>Device 1:<ul><li><a href="https://eu-central-1.saucelabs.com/jobs/c66f640878ad47198ae49fff03de9225?auth=97cd7472750f66e53448949f5b7dc0eb">Steps, video, logs</a></li><li><a href="https://assets.eu-central-1.saucelabs.com/jobs/c66f640878ad47198ae49fff03de9225/0018screenshot.png?auth=97cd7472750f66e53448949f5b7dc0eb">Failure screenshot</a></li></ul></p><p>Device 2:<ul><li><a href="https://eu-central-1.saucelabs.com/jobs/75d00ca9e54b4152bb12d76f7cb01f3e?auth=51fef4ef71872f0665ae04f950fb4408">Steps, video, logs</a></li><li><a href="https://assets.eu-central-1.saucelabs.com/jobs/75d00ca9e54b4152bb12d76f7cb01f3e/0011screenshot.png?auth=51fef4ef71872f0665ae04f950fb4408">Failure screenshot</a></li></ul></p></ul></p></td></tr></tbody></table></details><h3>Passed tests (2)</h3><details><summary>Click to expand</summary><br/><br/><h4>Class TestDeeplinkOneDevice:</h4><table style="width: 100%"><colgroup><col span="1" style="width: 20%;"><col span="1" style="width: 80%;"></colgroup><tbody><tr></tr><tr><td><b>1. <a href="https://ethstatus.testrail.net/index.php?/tests/view/3047166">test_deep_link_with_invalid_user_public_key_own_profile_key</a>, id: 702774 </b></td></tr><tr><td><ins>Device sessions</ins><p><ul><p>Device 1:<ul><li><a href="https://eu-central-1.saucelabs.com/jobs/11e17eb0207d44328cd7a142a18f8ab8?auth=9a43fdd0b4078a224b2f85ad76c1d855#40">Steps, video, logs</a></li></ul></p></ul></p></td></tr><tr><td><b>2. <a href="https://ethstatus.testrail.net/index.php?/tests/view/3047167">test_deep_link_open_user_profile</a>, id: 702775 </b></td></tr><tr><td><ins>Device sessions</ins><p><ul><p>Device 1:<ul><li><a href="https://eu-central-1.saucelabs.com/jobs/11e17eb0207d44328cd7a142a18f8ab8?auth=9a43fdd0b4078a224b2f85ad76c1d855#98">Steps, video, logs</a></li></ul></p></ul></p></td></tr></tbody></table></details>

churik commented:
@qfrank 
Can't open 1-1 chat on Android, that's the reason for failed e2e.
![IMAGE 2022-11-17 13:37:37](https://user-images.githubusercontent.com/4557972/202448434-a7645f70-2059-419e-b94d-4e39e98de136.jpg)
Works on IOS.

qfrank commented:
> @qfrank Can't open 1-1 chat on Android, that's the reason for failed e2e. ![IMAGE 2022-11-17 13:37:37](https://user-images.githubusercontent.com/4557972/202448434-a7645f70-2059-419e-b94d-4e39e98de136.jpg) Works on IOS.

Hi @churik , could u add more info on this ? I tried my android simulator and it works fine, thanks 
![image](https://user-images.githubusercontent.com/12406719/202452599-2f3caba8-ad3c-4953-b491-1bd83c0cabda.png)
![image](https://user-images.githubusercontent.com/12406719/202465751-e8f27267-5fc5-46ed-8d64-10877e0c1cc8.png)

churik commented:
sorry for not attaching logs from a real device, currently it is blocked by https://github.com/status-im/status-mobile/issues/14388 (seems to be fixed now)
But logs/video are available here

Video: https://app.eu-central-1.saucelabs.com/tests/c66f640878ad47198ae49fff03de9225?auth=97cd7472750f66e53448949f5b7dc0eb#54

Log: [logcat (3).log](https://github.com/status-im/status-mobile/files/10035653/logcat.3.log)

churik commented:
@qfrank thank you for the update!

On Xiaomi Mi 9T Pro (Android 10)  - there is no option for me at all:
![IMAGE 2022-11-18 14:54:24](https://user-images.githubusercontent.com/4557972/202720862-b0b89b8d-9c94-42cf-9692-702cf954f346.jpg)

On Samsung Galaxy A52 (Android 11) it can be opened, but not on the first attempt:
https://user-images.githubusercontent.com/4557972/202721185-3cb1ce12-61fb-426b-987e-5ed5b024836a.mp4

On Huawei P40 Lite, EMUI 12 (android 11): works as expected.

I can try some virtual devices if needed, let me know.

churik commented:
UPD: works fine on Samsung Galaxy with Android10, but not on Xiaomi phones.

qfrank commented:
i may need prepare a xiaomi device then :( , i will try it next week, thank you @churik

churik commented:
we have browserstack! I can share details

qfrank commented:
Hi @churik , this one just get fixed :)
> On Samsung Galaxy A52 (Android 11) it can be opened, but not on the first attempt:

churik commented:
I guess all the issues can be addressed separetely, not crucial from my POV.

VladimrLitvinenko commented:
@qfrank thank you for your work.
PR is tested and ready for merge. I have investigated WhatsApp on Xiaomi devices and there is the same behavior 
![image](https://user-images.githubusercontent.com/52490791/203373480-4c887f86-4d9f-448a-9f20-5ae95f456c06.png)

qfrank commented:
notice again, we need the IOS implementation :)
